### PR TITLE
Generic estimator support

### PR DESCRIPTION
**Pull request by jgoppert**
This closes https://github.com/mavlink/mavlink/issues/665.

### CODE REVIEW

**Comment by jgoppert** [comment]
@LorenzMeier @priseborough @bkueng  this should improve IEKF and EKF2 support for px4 and enable offboard slam coordination @mhkabir .

**Comment by jgoppert** [comment]
To support innovations we need something for each measurement, so it might be good to add these as well:

Extra MAV_FIELDS to add:
* VISION: x, y, z, q(0-4)
* MAG: body x, body y, body z, heading
* LIDAR: dist_bottom
* SONAR: dist_bottom
* GPS: pos_N, pos_E, asl, vel_N, vel_E, vel_D

Let me know if we need anything else.

**Comment by priseborough** [comment]
We also have FLOW: x,y (rad/sec)

Also I would prefer PosD instead of asl for all vertical position innovations. Our local coordinate frame is NED.

**Comment by jgoppert** [comment]
@priseborough OK, thanks, will add. I have both asl and PosD in there so that should be fine.

**Comment by jgoppert** [comment]
@priseborough The enum was getting to big, so I had to add an array indicating the sensor a field is associated to, so for instance:

id = [POS_N, POS_E, POS_D, BIAS_X, BIAS_Y, BIAS_Z, BIAS_X, BIAS_Y, BIAS_Z]
sensor = [NONE, NONE, NONE, GYRO, GYRO, GYRO, ACCEL, ACCEL, ACCEL]
state = [1, 1, 1, 1e-5, 1e-5, 1e-5, 1e-3, 1e-3, 1e-3]

meaning:
* pos_N = 1
* pos_E = 1
* pos_D = 1
* gyro bias x = 1e-5
* gyro bias y = 1e-5
* gyro bias z = 1e-5
* accel bias x = 1e-3
* accel bias y = 1e-3
* accel bias z = 1e-3

**Comment by jgoppert** [comment]
@priseborough @LorenzMeier @julianoes  I now have this working.

image

Just debugging why it isn't showing up in the analyze view of QGC now, any ideas?

**Comment by jgoppert** [comment]
Here is the patch that implements the sending above:

https://github.com/PX4/Firmware/pull/6045/commits/9da67ea11acceb2a6e120d6a14a3bf91aa4ae367

**Comment by jgoppert** [comment]
@tridge @LorenzMeier @priseborough  think this can get merged now?

**Comment by priseborough** [comment]
The optical flow is a delta angle (units of radians) but no delta time is provided.

**Comment by jgoppert** [comment]
@priseborough  So are you suggesting something like MAV_FIELD_DT, so you could encode type (MAV_FIELD_DT), sensor (OPTICAL_FLOW)? Would that be enough to cover your use case?

**Comment by priseborough** [comment]
Yes it would thanks.

**Comment by jgoppert** [comment]
@priseborough Look good to go now?

**Comment by priseborough** [comment]
Looks good to me.

**Comment by TSC21** [comment]
Is this to go forward? Really interested on this also. Thanks @jgoppert

**Comment by LorenzMeier** [comment]
@jgoppert Should we revisit and get this in?

**Comment by jgoppert** [comment]
I think so, it should help with diagnostics in the field,  @priseborough is this something you feel would still be useful for ekf2 as well?

**Comment by TSC21** [comment]
> I think so, it should help with diagnostics in the field, @priseborough is this something you feel would still be useful for ekf2 as well?

Very useful also for other estimators running offboard.

**Comment by priseborough** [comment]
Yes, for doing real time diagnostics it would be useful.

File: message_definitions/v1.0/common.xml [hunk 1]
```diff
@@ -4011,5 +4227,37 @@
       <field type="uint8_t" name="target_component">component ID of the target</field>
       <field type="uint16_t" name="sequence">sequence number (must match the one in LOGGING_DATA_ACKED)</field>
     </message>
+    <message id="269" name="ESTIMATOR_STATE">
```
Comment (julianoes) [inline message_definitions/v1.0/common.xml]: These message IDs are now taken. You need to move it 😄 .

Comment (TSC21) [reply message_definitions/v1.0/common.xml]: @jgoppert can you rebase and change the message ID?

File: message_definitions/v1.0/common.xml [hunk 2]
```diff
@@ -2434,6 +2434,222 @@
...
+      </entry>
+      <entry value="8" name="MAV_SENSOR_TYPE_PITOT">
+        <description>Pitot tube</description>
+      </entry>
+      <entry value="9" name="MAV_SENSOR_TYPE_MOCAP">
+        <description>Motion capture</description>
+      </entry>
+      <entry value="10" name="MAV_SENSOR_TYPE_FLOW">
+        <description>Optical flow</description>
+      </entry>
```
Comment (TSC21) [inline message_definitions/v1.0/common.xml]: We may want to add `VISION` also?

**Comment by julianoes** [comment]
Implemented by #873.

**Pull request closed without merging**